iOS/ViewControllers/Account/DownloadListViewController.swift

```diff
@@ -0,0 +1,345 @@
+        let videoRequest = VideoHelper.FetchRequest.hasDownloadedVideo()
+        var items: [DownloadItem] = []
+        let promise = Promise<[DownloadItem], XikoloError>()
+        CoreDataHelper.persistentContainer.performBackgroundTask { privateManagedObjectContext in
+            do {
+                let downloadedVideos = try privateManagedObjectContext.fetch(videoRequest)
+                for video in downloadedVideos {
+                    if let course = video.item?.section?.course {
+                        items.append(DownloadItem(courseID: course.id, courseTitle: course.title, contentType: .video))
+                    }
+                }
+
+                return promise.success(items)
+            } catch {
+                promise.failure(.coreData(error))
+            }
+        }
+
+        return promise.future
+    }
+
+    func getSlidesCourseIDs() -> Future<[DownloadItem], XikoloError> {
+        let slidesRequest = VideoHelper.FetchRequest.hasDownloadedSlides()
+        var items: [DownloadItem] = []
+        let promise = Promise<[DownloadItem], XikoloError>()
+        CoreDataHelper.persistentContainer.performBackgroundTask { privateManagedObjectContext in
+            do {
+                let downloadedSlides = try privateManagedObjectContext.fetch(slidesRequest)
+                for slide in downloadedSlides {
+                    if let course = slide.item?.section?.course {
+                        items.append(DownloadItem(courseID: course.id, courseTitle: course.title, contentType: .slides))
+                    }
+                }
+
+                return promise.success(items)
+            } catch {
+                promise.failure(.coreData(error))
+            }
+        }
+
+        return promise.future
+    }
+
+    func getDocumentsCourseIDs() -> Future<[DownloadItem], XikoloError> {
+        let documentsRequest = DocumentHelper.FetchRequest.hasDownloadedLocalization()
+        var items: [DownloadItem] = []
+        let promise = Promise<[DownloadItem], XikoloError>()
+        CoreDataHelper.persistentContainer.performBackgroundTask { privateManagedObjectContext in
+            do {
+                let downloadedDocuments = try privateManagedObjectContext.fetch(documentsRequest)
+                if !downloadedDocuments.isEmpty {
+                    self.hasDocuments = true
+                }
+
+                for document in downloadedDocuments {
+                    let downloadItems = document.courses.map { course -> DownloadItem in
+                        return DownloadItem(courseID: course.id, courseTitle: course.title, contentType: .document)
+                    }
+
+                    items.append(contentsOf: downloadItems)
+                }
+
+                return promise.success(items)
+            } catch {
+                promise.failure(.coreData(error))
+            }
+        }
+
+        return promise.future
+    }
```

Comment:
As far as I can see, these these methods `getVideoCourseIDs`, `getSlidesCourseIDs` and `getDocumentsCourseIDs` are pretty similar. They only differ in the fetch request, the content type of `DownloadItem` and the mapping from resource to course (e.g. documents to course). I think here we can make use of a generic function with take these as parameters. 

We can to this together when you are in the office the next time 😃